docs/proposals/configmap-rollout.md

```diff
@@ -0,0 +1,171 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# ConfigMap management
+
+## Abstract
+
+A proposal for implementing management of ConfigMaps in the context of a
+[Deployment](http://kubernetes.io/docs/user-guide/deployments/).
+
+Users should be able to have their apps automatically pick up changes to
+their configuration and also have the ability to rollback to previous
+configuration similar to the ability to rollback to previous pod template
+revisions.
+
+Currently, in order to apply updated configuration into a running app, users
+need to 1) either do it manually: copy the ConfigMap that holds their old
```

Comment:
We need to be really careful about the relationship here.  A configmap is unlikely to be owned by a deployment in the classic sense (a primary configmap - the one that an admin defines that contains config that is used by many components).  It's very likely instead that the configmap is intended to flow out to deployments, and so the deployments are subordinate to the configmap.  That is why I don't think ownerRef is appropriate on the deployment.

I think it's more useful to think about this the other direction - "how do I flow configmap/secret changes out to 1..N entities" and preserve safe updates.  That's the underlying use case, and so the steps need to reflect it.  I want to change my config declaratively.  The question about whether it should be automated (whether the complexity is worth it) is 

a) how many times would an average user do this (change the config)
b) is there a better, simpler flow (apply from a checked in config)
c) is that better simpler flow "safer"?

I think my worry with the proposed 1 is that it is not safe - a naive user is likely to change the configmap and get burned when their app magically updates without a deployment.  Can we better direct users to handle this?

Also, with apply+purge there's another possibility for 1 being messed up - the user renames the config map in their config directory from config-1 to config-2, which means apply+purge deletes config-1, which means that pods in the old deployment are now broken.  The user has to leave config-1 in their source config until it is no longer used (which is a hard determination).  A particular version of a configmap is something that needs to be available and immutable until it is no longer referenced by a pod, and I can't see a way to do that easily today.  We would need ownerRef on that config to be owned by an instance of the deployment, not by the deployment.